filters/include/pcl/filters/experimental/voxel_grid.h

```diff
@@ -0,0 +1,513 @@
+        std::floor((max_p[1] - min_p[1]) * inverse_leaf_size_[1]) + 1;
+    const std::size_t dz =
+        std::floor((max_p[2] - min_p[2]) * inverse_leaf_size_[2]) + 1;
+    const size_t max_size = std::numeric_limits<std::size_t>::max();
+
+    // check unsigned int overflow/wrap-around
+    if (dy > max_size / dx || dx * dy > max_size / dz)
+      return boost::none;
+    else
+      return dx * dy * dz;
+  }
+
+  inline std::size_t
+  hashPoint(const PointT& pt)
+  {
+    const std::size_t ijk0 = std::floor(pt.x * inverse_leaf_size_[0]) - min_b_[0];
+    const std::size_t ijk1 = std::floor(pt.y * inverse_leaf_size_[1]) - min_b_[1];
+    const std::size_t ijk2 = std::floor(pt.z * inverse_leaf_size_[2]) - min_b_[2];
+    return ijk0 * divb_mul_[0] + ijk1 * divb_mul_[1] + ijk2 * divb_mul_[2];
+  }
+
+  inline void
+  addPointToGrid(const PointT& pt)
+  {
+    if (!filter_field_name_.empty()) {
+      // Get the distance value
+      const std::uint8_t* pt_data = reinterpret_cast<const std::uint8_t*>(&pt);
+      float distance_value = 0;
+      memcpy(&distance_value,
+             pt_data + filter_fields_[filter_field_idx_].offset,
+             sizeof(float));
+
+      if (getDerived()->getFilterLimitsNegative()) {
+        // Use a threshold for cutting out points which inside the interval
+        if ((distance_value < filter_limit_max_) &&
+            (distance_value > filter_limit_min_))
+          return;
+      }
+      else {
+        // Use a threshold for cutting out points which are too close/far away
+        if ((distance_value > filter_limit_max_) ||
+            (distance_value < filter_limit_min_))
+          return;
+      }
+    }
+
+    const std::size_t h = hashPoint(pt);
+    grid_[h].add(pt);
+  }
+
+  inline boost::optional<PointT>
+  filterGrid(GridIterator grid_it)
+  {
+    auto& voxel = grid_it->second;
+    const std::size_t min_points_per_voxel =
+        getDerived()->getMinimumPointsNumberPerVoxel();
+    if (voxel.getSize() >= min_points_per_voxel) {
+
+      const bool save_leaf_layout = getDerived()->getSaveLeafLayout();
+      if (save_leaf_layout)
+        leaf_layout_[grid_it->first] = num_voxels_++;
+
+      PointT centroid;
+      const bool downsample_all_data = getDerived()->getDownsampleAllData();
+      if (!downsample_all_data) {
+        // Extract coordinate only
+        PointT pt;
+        voxel.get(pt);
+        centroid.getVector4fMap() = pt.getVector4fMap();
+      }
+      else {
+        voxel.get(centroid);
+      }
+      return centroid;
+    }
+
+    return boost::none;
+  }
+
+protected:
+  /** \brief The filter name. */
+  std::string filter_name_;
+
+  /** \brief The size of a leaf. */
+  Eigen::Vector4f leaf_size_;
+
+  /** \brief Internal leaf sizes stored as 1/leaf_size_ for efficiency reasons. */
+  Eigen::Array4f inverse_leaf_size_;
+
+  /** \brief The leaf layout information for fast access to cells relative to current
+   * position **/
+  std::vector<int> leaf_layout_;
```

Comment:
`vector<int>`  ->

`vector<size_t>` OR `unordered_map<size_t,size_t>` ?

Reply:
`vector<size_t>` if we maintain compatibility, else `unordered_map<size_t, size_t>`